man/detroit-yard.5
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "DETROIT\-YARD" "5" "October 2011" "" ""
.
.SH "NAME"
\fBdetroit\-yard\fR \- generate yard api documentation via detroit
.
.SH "DESCRIPTION"
The YARD plug\-in for Detroit utilizes the \fByard\fR documentation tool to generate API documentation for a project\. It will also generate yri documentation files\.
.
.SH "OPTIONS"
The following options can be used in the Detroit assembly file to define a YARD service\.
.
.IP "\(bu" 4
\fByardopts\fR \- If set to true, use \fB\.yardopts\fR file and ignore other settings\. This defaults to \fBtrue\fR if no other settings are given and a \fB\.yardopts\fR file exists\.
.
.IP "\(bu" 4
\fBtitle\fR \- Title of documents\. Defaults to general metadata title field\.
.
.IP "\(bu" 4
\fBoutput\fR \- Where to save generated documentation\. The default is \fBdoc/\fR\.
.
.IP "\(bu" 4
\fBtemplate\fR \- Template to use\. This defaults to \fBENV[\'YARD_TEMPLATE\']\fR\.
.
.IP "\(bu" 4
\fBfiles\fR/\fBinclude\fR \- Which files to document\. The default is \fBlib/\fR\.
.
.IP "\(bu" 4
\fBexclude\fR \- Paths to specifically exclude\.
.
.IP "\(bu" 4
\fBignore\fR \- File patterns to ignore\.
.
.IP "\(bu" 4
\fBdocs\fR \- Which project document files to include, e\.g\. \fBCOPYING\fR\.
.
.IP "\(bu" 4
\fBreadme\fR \- Main file\. This can be a file pattern\. This default is matched to \fBREADME{,\.*}\fR\.
.
.IP "\(bu" 4
\fBextra\fR \- Additional options passed to the \fBrdoc\fR command\.
.
.IP "" 0
.
.SH "EXAMPLES"
Usually \fB\.yardopts\fR is used, in which case the simplest entry would be:
.
.IP "" 4
.
.nf

yard:
  priority: 1
.
.fi
.
.IP "" 0
.
.P
We set the priority lower than the default (highest is \fB0\fR) in case some other documentation tool generates files that should be included in the YARD docs\.
.
.P
A more complex example might look like:
.
.IP "" 4
.
.nf

yard:
  title: MyApp API
  readme: INTRO\.rdoc
  include:
     \- lib
  docs:
     \- [A\-Z]*\.*
  ignore:
    \- \'*\.yml\'
  output: site/api
  priority: 1
.
.fi
.
.IP "" 0
.
.P
The above examples use the YAML\-base assembly format\. Assembly files can also be Ruby\-based\. See Detroit documentation for more details\.
.
.SH "RESOURCES"
.
.IP "\(bu" 4
API Documentation \fIhttp://rubydoc\.info/gems/detroit\-yard\fR
.
.IP "\(bu" 4
Development Site \fIhttp://github\.com/detroit/detroit\-yard\fR
.
.IP "" 0
.
.SH "COPYRIGHT"
Copyright (c) 2010 Thomas Sawyer, Rubyworks
.
.P
Detroit YARD is distributable in accordance with the GPL v\.3 license\.
.
.SH "SEE ALSO"
detroit(1), yard(1)


lib/detroit-yard.rb
require 'detroit/tool'

module Detroit

  # Yard tool.
  def Yard(options={})
    Yard.new(options)
  end

  # Yard documentation tool generates YARD documentation for a project.
  #
  # By default it places the documentaiton file in the standard `doc`
  # directory unless a `site/yard` or `yard` directory exists, in which
  # case the documentation will be stored there.
  #
  #--
  # TODO: Should this autodetect .yardopts and use them unless told
  # to do otherwise?
  #
  # TODO: Not sure the #current? code is exactly correct, might get
  # false negatives.
  #++
  class Yard < Tool

    # Default location to store yard documentation files.
    DEFAULT_OUTPUT = "doc"

    # Locations to check for existance in deciding where to store
    # yard documentation.
    DEFAULT_OUTPUT_MATCH = "{site/,website/,doc/}yard"

    # Default main file.
    DEFAULT_README = "README"

    # Default template to use.
    DEFAULT_TEMPLATE = "default"

    # Deafult extra options to add to yardoc call.
    DEFAULT_EXTRA = ''

    #
    #DEFAULT_FILES = 'lib/**/*;bin/*'

    #
    #DEFAULT_TOPFILES = '[A-Z]*'


    #  A T T R I B U T E S

    # If set to true, use `.yardopts` file and ignore other settings.
    attr_accessor :yardopts

    # Title of documents. Defaults to general metadata title field.
    attr_accessor :title

    # Directory in which to save yard files.
    attr_accessor :output

    # Template to use (defaults to ENV['YARD_TEMPLATE'] or 'default')
    attr_accessor :template

    # Main file.  This can be file pattern. (README{,.txt})
    attr_accessor :readme

    # Which library files to document.
    attr_reader :files

    #
    def files=(list)
      @resolved_files = nil
      @files = list.to_list
    end

    # More generic alternative to +files+.
    alias_accessor :include, :files

    # Which project top-files to document.
    attr_reader :topfiles

    # Set topfiles list.
    def topfiles=(list)
      @resolved_topfiles = nil
      @topfiles = list.to_list
    end

    # Alias for +topfiles+.
    alias_accessor :docs, :topfiles

    # Paths to specifically exclude.
    attr_accessor :exclude

    # File patterns to ignore.
    attr_accessor :ignore

    # Additional options passed to the yardoc command.
    attr_accessor :extra


    #  A S S E M B L Y  S T A T I O N S

    # Attach document method to assembly.
    def station_document
      document
    end

    # Attach reset method to assembly.
    def station_reset
      reset
    end

    # Attach clean method to assembly.
    def station_clean
      clean
    end

    # Attach purge method to assembly.
    def station_purge
      purge
    end


    #  M E T H O D S

    # Are YARD docs current and not in need of updating?
    # If yes, returns string message, otherwise `false`.
    def current?
      if outofdate?(output, *(resolved_files + resolved_topfiles))
        false
      else
        "YARD docs are current (#{output})."
      end
    end

    # Generate documentation. Settings are the
    # same as the yardoc command's option, with two
    # exceptions: +inline+ for +inline-source+ and
    # +output+ for +op+.
    def document
      title    = self.title
      output   = self.output
      readme   = self.readme
      template = self.template
      #exclude  = self.exclude
      extra    = self.extra

      # TODO: add to resolved_topfiles ?
      readme = Dir.glob(readme, File::FNM_CASEFOLD).first

      if (msg = current?) && ! force?
        report msg
      else
        if !yardopts
          status "Generating YARD documentation in #{output}."
        end

        #target_main = Dir.glob(target['main'].to_s, File::FNM_CASEFOLD).first
        #target_main   = File.expand_path(target_main) if target_main
        #target_output = File.expand_path(File.join(output, subdir))
        #target_output = File.join(output, subdir)

        if yardopts
          argv = []
        else
          argv = []
          argv.concat(String === extra ? extra.split(/\s+/) : extra)
          argv.concat ['--output-dir', output] if output
          argv.concat ['--readme', readme] if readme
          argv.concat ['--template', template] if template
          argv.concat ['--title', title] if title
          #argv.concat ['--exclude', exclude]
          argv.concat resolved_files
          argv.concat ['-', *resolved_topfiles]
        end

        yard_target(output, argv)

        touch(output) if File.directory?(output) unless yardopts
      end
    end

    # Mark the output directory as out of date.
    def reset
      if directory?(output) && !yardopts
        utime(0, 0, output)
        report "Reset #{output}" #unless trial?
      end
    end

    # TODO: remove .yardoc ?
    def clean
    end

    # Remove yardoc output directory.
    def purge
      if directory?(output)
        rm_r(output)
        status "Removed #{output}" unless trial?
      end
    end

  private

    #
    def initialize_defaults
      @title    = metadata.title
      @files    = metadata.loadpath + ['bin'] # DEFAULT_FILES
      @topfiles = ['[A-Z]*']

      @output   = Dir[DEFAULT_OUTPUT_MATCH].first || DEFAULT_OUTPUT
      @readme   = DEFAULT_README
      @extra    = DEFAULT_EXTRA
      @template = ENV['YARD_TEMPLATE'] || DEFAULT_TEMPLATE
    end

    #
    LOCAL_SETTINGS = %w{files includes topfiles docs readme output template exclude ignore extra}

    # If there are no options set than default to using the .yardopts file (if it exists).
    def initialize_options(options)
      if !options.keys.any?{ |k| LOCAL_SETTINGS.include?(k) }
        @yardopts = true if (project.root + '.yardopts').exist?
      end
      super(options)
    end

    #
    def resolved_files
      @resolved_files ||= (
        amass(files, exclude || [], ignore || []).uniq
      )
    end

    #
    def resolved_topfiles
      @resolved_topfiles ||= (
        amass(topfiles, exclude || [], ignore || [])
      )
    end

    # Generate yardocs for input targets.
    def yard_target(output, argv=[])
      # remove old yardocs
      #rm_r(output) if exist?(output) and safe?(output)
      #options['output-dir'] = output
      #args = "#{extra} " + [input, options].to_console
      #argv = args.split(/\s+/)

      args = argv.join(' ')
      cmd  = "yardoc " + args

      trace(cmd)

      if trial?
      else
        YARD::CLI::Yardoc.run(*argv)
      end
    end

    # Require yard library.
    def initialize_requires
      require 'yard'
    end

  public

    def self.man_page
      File.dirname(__FILE__)+'/../man/detroit-yard.5'
    end

  end

end
